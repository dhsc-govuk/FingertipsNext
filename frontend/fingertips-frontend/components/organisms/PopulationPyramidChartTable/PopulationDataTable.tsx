'use client';

import {
  PopulationDataForArea,
  getLowerBandValue,
} from '@/lib/chartHelpers/preparePopulationData';
import { Table } from 'govuk-react';
import { typography } from '@govuk-react/lib';
import { StyledAlignLeftHeader } from '@/lib/tableHelpers';
import styled from 'styled-components';
import { formatWholeNumber } from '@/lib/numberFormatter';

const StyledAreaTitleHeader = styled('h3')(typography.font({ size: 19 }), {
  textAlign: 'center',
  height: '65px',
  lineHeight: '65px',
  letterSpacing: '0px',
  fontWeight: '700',
  padding: '5px',
});

export const StyledTableCell = styled(Table.Cell)(
  typography.font({ size: 19 }),
  {
    paddingRight: '0',
    textAlign: 'center',
  }
);

const valueFormatter = (value: number | string | undefined) => {
  if (typeof value === 'number' || typeof value === 'undefined') {
    return formatWholeNumber(value);
  }
  return value;
};

const StyledAreaNameHeader = styled(StyledAlignLeftHeader)(
  typography.font({ size: 19 }),
  {
    borderTop: `solid #F3F2F1 2px`,
    textAlign: 'center',
    fontWeight: 700,
  }
);

const getSortAgeBandIndexes = (ageBands: string[] | undefined): number[] => {
  if (!ageBands) return [];

  return ageBands
    .map((item, index) => ({ index, value: item }))
    .sort((a, b) => {
      return getLowerBandValue(a.value) - getLowerBandValue(b.value);
    })
    .map((item) => item.index);
};

type ItemDataType = string | undefined | number;

interface PopulationDataTableProps {
  headers: string[];
  title: string;
  healthDataForArea: PopulationDataForArea | undefined;
  filterValues?: (header: ItemDataType[]) => ItemDataType[];
}

const computeFooterItems = (
  maleSeries: (number | undefined)[],
  femaleSeries: (number | undefined)[]
): ItemDataType[] => {
  const males = maleSeries.reduce((prev, current) => {
    return (prev ?? 0) + (current ?? 0);
  }, 0);

  const females = femaleSeries.reduce((prev, current) => {
    return (prev ?? 0) + (current ?? 0);
  }, 0);

  return ['All ages', males, females];
};

export const PopulationDataTable = ({
  headers,
  healthDataForArea,
  title,
  filterValues,
}: PopulationDataTableProps) => {
  if (!healthDataForArea) return <></>;

  const indexes = getSortAgeBandIndexes(healthDataForArea?.ageCategories);

  const points = healthDataForArea.ageCategories.map((value, index) => [
    value,
    healthDataForArea.femaleSeries[index],
    healthDataForArea.maleSeries[index],
  ]);

  const footerRowItems = computeFooterItems(
    healthDataForArea.maleSeries,
    healthDataForArea.femaleSeries
  );
  const filterFooterRowItems = filterValues
    ? filterValues(footerRowItems)
    : footerRowItems;
  return (
    <section>
      <StyledAreaTitleHeader>{title}</StyledAreaTitleHeader>

      <Table
        head={
          <Table.Row>
            {headers.map((header) => {
              return (
                <StyledAreaNameHeader key={header}>
                  {header}
                </StyledAreaNameHeader>
              );
            })}
          </Table.Row>
        }
      >
        {indexes.map((index) => {
          const columnValues = filterValues
            ? filterValues(points[index])
            : points[index];

          return (
            <Table.Row key={`${healthDataForArea.areaName}-${index}`}>
              {columnValues.map((value, valueIndex: number) => (
                <StyledTableCell
                  key={`${healthDataForArea.areaName}-${index}-${valueIndex}`}
                >
                  {valueFormatter(value)}
                </StyledTableCell>
              ))}
            </Table.Row>
          );
        })}
        <Table.Row key={`footer-${healthDataForArea.areaName}`}>
          {filterFooterRowItems.map((value, valueIndex: number) => (
            <StyledAreaNameHeader
              key={`footer-${healthDataForArea.areaName}-${valueIndex}`}
            >
              {valueFormatter(value)}
            </StyledAreaNameHeader>
          ))}
        </Table.Row>
      </Table>
    </section>
  );
};
