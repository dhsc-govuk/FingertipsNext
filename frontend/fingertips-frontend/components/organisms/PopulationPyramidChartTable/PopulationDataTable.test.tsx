import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { PopulationDataTable } from './PopulationDataTable';

const mockHealthDataForArea = {
  areaName: 'Test Area',
  ageCategories: ['0-4', '5-9', '10-14', '15+'],
  femaleSeries: [1000, 1200, 1100, 1300],
  maleSeries: [900, 1100, 1050, 1250],
  total: 0,
};

describe('PopulationDataTable', () => {
  test('renders table with correct headers and data', () => {
    render(
      <PopulationDataTable
        headers={['Age Band', 'Females', 'Males']}
        title="Population Data"
        healthDataForArea={mockHealthDataForArea}
      />
    );

    expect(screen.getByText('Population Data')).toBeInTheDocument();
    expect(screen.getByText('Age Band')).toBeInTheDocument();
    expect(screen.getByText('Females')).toBeInTheDocument();
    expect(screen.getByText('Males')).toBeInTheDocument();
    expect(screen.getByText('0-4')).toBeInTheDocument();
    expect(screen.getByText('1,000')).toBeInTheDocument();
    expect(screen.getByText('900')).toBeInTheDocument();
  });

  test('renders empty when no data is provided', () => {
    const { container } = render(
      <PopulationDataTable
        headers={['Age Band', 'Females', 'Males']}
        title="Population Data"
        healthDataForArea={undefined}
      />
    );
    expect(container.firstChild).toBeNull();
  });

  test('take a snapshot', () => {
    const { asFragment } = render(
      <PopulationDataTable
        headers={['Age Band', 'Females', 'Males']}
        title="Population Data"
        healthDataForArea={mockHealthDataForArea}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
