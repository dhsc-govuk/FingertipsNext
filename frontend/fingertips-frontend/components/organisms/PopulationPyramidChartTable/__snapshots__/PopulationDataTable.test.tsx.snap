// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopulationDataTable take a snapshot 1`] = `
<DocumentFragment>
  .c2 {
  padding: 10px 20px 10px 0;
  border-bottom: 1px solid #bfc1c3;
  text-align: left;
  font-weight: 700;
}

.c2:last-child {
  padding-right: 0;
}

.c6 {
  padding: 10px 20px 10px 0;
  border-bottom: 1px solid #bfc1c3;
  text-align: left;
}

.c6:last-child {
  padding-right: 0;
}

.c1 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  color: #0b0c0c;
  width: 100%;
  border-spacing: 0;
  border-collapse: collapse;
  margin-bottom: 20px;
}

.c3 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  font-weight: bold;
  padding: 0.625em 0;
}

.c7 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  padding-right: 0;
}

.c4 {
  text-align: left;
  vertical-align: top;
}

.c0 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  text-align: center;
  height: 65px;
  line-height: 65px;
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  font-weight: 700;
}

.c5 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  border-top: solid #F3F2F1 2px;
  text-align: center;
  line-height: 1.31579;
  font-weight: 700;
}

@media print {
  .c1 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c1 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c1 {
    color: #000;
  }
}

@media only screen and (min-width:641px) {
  .c1 {
    margin-bottom: 30px;
  }
}

@media print {
  .c3 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c7 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c7 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c0 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c0 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c5 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c5 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

<section>
    <h3
      class="c0"
    >
      Population Data
    </h3>
    <table
      class="c1"
    >
      <thead
        class=""
      >
        <tr
          class=""
        >
          <th
            class="c2 c3 c4 c5"
          >
            Age Band
          </th>
          <th
            class="c2 c3 c4 c5"
          >
            Females
          </th>
          <th
            class="c2 c3 c4 c5"
          >
            Males
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      >
        <tr
          class=""
        >
          <td
            class="c6 c7"
          >
            0-4
          </td>
          <td
            class="c6 c7"
          >
            900
          </td>
          <td
            class="c6 c7"
          >
            1,000
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class="c6 c7"
          >
            5-9
          </td>
          <td
            class="c6 c7"
          >
            1,100
          </td>
          <td
            class="c6 c7"
          >
            1,200
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class="c6 c7"
          >
            10-14
          </td>
          <td
            class="c6 c7"
          >
            1,050
          </td>
          <td
            class="c6 c7"
          >
            1,100
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class="c6 c7"
          >
            15+
          </td>
          <td
            class="c6 c7"
          >
            1,250
          </td>
          <td
            class="c6 c7"
          >
            1,300
          </td>
        </tr>
        <tr
          class=""
        >
          <th
            class="c2 c3 c4 c5"
          >
            All ages
          </th>
          <th
            class="c2 c3 c4 c5"
          >
            4,300
          </th>
          <th
            class="c2 c3 c4 c5"
          >
            4,600
          </th>
        </tr>
      </tbody>
    </table>
  </section>
</DocumentFragment>
`;
